When using gcc, use '-pthread' (or '-pthreads' on solaris)
compiler flag to do all pthread specific things.
--- configure.in.orig	2009-04-02 17:25:45 +0200
+++ configure.in	2009-04-02 17:32:00 +0200
@@ -524,7 +524,19 @@
       )
       AC_CHECK_HEADERS(semaphore.h)
       AC_DEFINE(HAVE_PTHREADS)
+      case "${GXX}:${target}" in
+      yes:*solaris*|yes:*sunos*)
+        CXXFLAGS="-pthreads ${CXXFLAGS}"
+        LDFLAGS="-pthreads ${LDFLAGS}"
+	;;
+      yes:*)
+	CXXFLAGS="-pthread ${CXXFLAGS}"
+	LDFLAGS="-pthread ${LDFLAGS}"
+	;;
+      *)
       SHLIBS="-lpthread"
+      ;;
+      esac
     else 
       AC_CHECK_LIB(c_r, pthread_create, use_pthreads=yes, use_pthreads=no)
       if test X"$use_pthreads" = Xyes; then
--- configure.orig	2009-04-02 17:25:42 +0200
+++ configure	2009-04-02 17:32:00 +0200
@@ -2738,7 +2738,19 @@
 #define HAVE_PTHREADS 1
 EOF
 
+      case "${GXX}:${target}" in
+      yes:*solaris*|yes:*sunos*)
+        CXXFLAGS="-pthreads ${CXXFLAGS}"
+        LDFLAGS="-pthreads ${LDFLAGS}"
+	;;
+      yes:*)
+	CXXFLAGS="-pthread ${CXXFLAGS}"
+	LDFLAGS="-pthread ${LDFLAGS}"
+	;;
+      *)
       SHLIBS="-lpthread"
+        ;;
+      esac
     else 
       echo $ac_n "checking for pthread_create in -lc_r""... $ac_c" 1>&6
 echo "configure:2745: checking for pthread_create in -lc_r" >&5
